File: IPWorks EDI Translator Samples/EDIFACT Translator/c++/edifacttranslator.cpp
/*
 * IPWorks EDI Translator 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks EDI Translator in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworkseditranslator
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworkseditranslator.h"
#define LINE_LEN 100

char input[LINE_LEN + 1];
const char* prompt(const char* prompt, const char* defaultVal) {
  printf("%s [%s]: ", prompt, defaultVal);
  fgets(input, LINE_LEN, stdin);
  input[strlen(input) - 1] = '\0';
  if (strlen(input) == 0) {
    strncpy(input, defaultVal, LINE_LEN);
    input[LINE_LEN] = '\0';
  }
  return input;
}
const bool yesno(const char* prompt, const bool defaultVal) {
  char def = defaultVal ? 'y' : 'n';
  printf("%s (y/n) [%c]: ", prompt, def);
  fgets(input, LINE_LEN, stdin);
  input[strlen(input) - 1] = '\0';
  if (strlen(input) == 0) return defaultVal;
  char answer = input[0];
  if (answer == 'y' || answer == 'Y') return true;
  else if (answer == 'n' || answer == 'N') return false;
  else return defaultVal;
}

int main(int argc, char *argv[])
{
	EDIFACTTranslator ediTranslator;

	printf("****************************************************************\n");
	printf("* This demo shows how to use the EDIFACTTranslator component to    *\n");
	printf("* convert an EDI document to XML, or vice versa. A sample EDI  *\n");
	printf("* document and corresponding schema is provided.               *\n");
	printf("****************************************************************\n");

    switch (atoi(prompt("Possible input format values are:\n0 - XML\n1 - JSON\n2 - EDIFACT\nPlease select a value", "2"))) {
		case 0:
			ediTranslator.SetInputFormat(EIF_XML);
			break;
		case 1:
			ediTranslator.SetInputFormat(EIF_JSON);
			break;
		case 2:
			ediTranslator.SetInputFormat(EIF_EDIFACT);
			break;
	}
	
	if (ediTranslator.GetInputFormat() == 0) {
		// InputFormat is XML, so OutputFormat must be an EDI format.
		switch (atoi(prompt("Possible output format values are:\n0 - XML\n1 - JSON\n2 - EDIFACT\nPlease select a value", ""))) {
			case 0:
				ediTranslator.SetOutputFormat(EOF_XML);
				break;
			case 1:
				ediTranslator.SetOutputFormat(EOF_JSON);
				break;
			case 2:
				ediTranslator.SetOutputFormat(EOF_EDIFACT);
				break;
		}
	} else {
		// InputFormat is EDI, so OutputFormat must be XML.
		printf("Output format set to XML.\n");
		ediTranslator.SetOutputFormat(EOF_XML);
	}

	if (ediTranslator.GetInputFormat() > 1) {
		if (yesno("Do you want to load a schema file?", true)) {
			ediTranslator.LoadSchema(prompt("Specify schema file path", (ediTranslator.GetInputFormat() == 2 ? "./RSSBus_D97A_INVOIC.json" : "")));
		}
	}

	ediTranslator.SetInputFile(prompt("Specify input file to translate", (ediTranslator.GetInputFormat() == 2 ? "./INVOIC.edi" : "")));

	ediTranslator.SetOutputFile(prompt("Specify output file", (ediTranslator.GetOutputFormat() > 1 ? "./output.edi" : "./output.xml")));
	
	ediTranslator.SetOverwrite(true);
	int ret_code = ediTranslator.Translate();

	if (ret_code)
		printf("ERROR:%i:%s\n", ret_code, ediTranslator.GetLastError());
	else
		printf("Translation Complete\n");

	printf("\nPress any key to continue.");
    getchar();
	return 0;
}

File: IPWorks EDI Translator Samples/EDIFACT Translator/net/edifacttranslator.cs
/*
 * IPWorks EDI Translator 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks EDI Translator in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworkseditranslator
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using nsoftware.IPWorksEDITranslator;

class edifacttranslatorDemo
{
  private static EDIFACTTranslator edifacttranslator = new nsoftware.IPWorksEDITranslator.EDIFACTTranslator();

  static void Main(string[] args)
  {
    if (args.Length < 6)
    {
      Console.WriteLine("usage: edifacttranslator /tm translationmode /f inputfile /o outputfile");
      Console.WriteLine("  translationmode  1 to translate EDI to XML or 2 to translate XML to EDI");
      Console.WriteLine("  inputfile        the file to translate");
      Console.WriteLine("  outputfile       the file to which the translated data will be written\n");
      Console.WriteLine("Example: edifacttranslator /tm 1 /f ./INVOIC.edi /o ./INVOICE.xml");
      Console.WriteLine("Example: edifacttranslator /tm 2 /f ./INVOICE.xml /o ./INVOIC_fromxml.edi");
    }
    else
    {
      try
      {
        System.Collections.Generic.Dictionary<string, string> myArgs = ConsoleDemo.ParseArgs(args);

        switch (myArgs["tm"])
        {
          case "1":
            Console.WriteLine("Converting EDIFACT File to XML File.");
            edifacttranslator.Reset();
            edifacttranslator.LoadSchema("../../../RSSBus_D97A_INVOIC.json");
            edifacttranslator.InputFile = myArgs["f"];
            edifacttranslator.OutputFile = myArgs["o"];
            edifacttranslator.Overwrite = true;

            edifacttranslator.Translate();
            Console.WriteLine("Translation complete.");
            break;
          case "2":
            Console.WriteLine("Converting XML File to EDIFACT File.");
            edifacttranslator.Reset();
            edifacttranslator.InputFormat = EDIFACTTranslatorInputFormats.eifXML;
            edifacttranslator.OutputFormat = EDIFACTTranslatorOutputFormats.eofEDIFACT;
            edifacttranslator.InputFile = myArgs["f"];
            edifacttranslator.OutputFile = myArgs["o"];
            edifacttranslator.Overwrite = true;

            edifacttranslator.Translate();
            Console.WriteLine("Translation complete.");
            break;
          default:
            Console.WriteLine("Unrecognized option.");
            break;
        }
      }
      catch (Exception ex)
      {
        Console.WriteLine(ex.Message);
      }
    }
  }
}




class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add an key to the dictionary for each argument
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/" then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}